src/browser.c

```diff
@@ -0,0 +1,147 @@
+   if (stat(path, &statbuf) != 0)
+       return 0;
+   return S_ISDIR(statbuf.st_mode);
+}
+
+void load_directory(browser* context, const char *dir) {
+    FTS* file_system = NULL;
+    FTSENT* child = NULL;
+    FTSENT* parent = NULL;
+
+    file_system = fts_open((char* const*)&dir, FTS_COMFOLLOW | FTS_NOCHDIR, &compare);
+
+    if (file_system) {
+        while( (parent = fts_read(file_system))) {
+            child = fts_children(file_system, 0);
+            if (errno != 0) {
+                fprintf(stderr, "Unable to load directory %s: %s\n", dir, strerror(errno));
+            }
+
+            char* file = malloc(PATH_MAX);
+            while (child && child->fts_link) {
+                child = child->fts_link;
+                if (child->fts_info == FTS_F && (context->recursive || child->fts_level == 1)) {
+                    snprintf(file, PATH_MAX, "%s%s", child->fts_path, child->fts_name);
+                    add_file(context, file);
+                }
+            }
+            free(file);
+        }
+        fts_close(file_system);
+    }
+}
+
+browser* create_browser(const char** paths, size_t paths_num, bool recursive)
+{
+    browser* browser = malloc(sizeof(browser));
+    if (!browser) {
+        fprintf(stderr, "Not enough memory\n");
+        return NULL;
+    }
+    browser->max = 1024;
+    browser->total = 0;
+    browser->current = -1;
+    browser->recursive = recursive;
+    browser->files = (char**)malloc(browser->max * sizeof(char*));
+    if (!browser->files) {
+        fprintf(stderr, "Not enough memory\n");
+        return NULL;
+    }
+
+    if (paths_num > 0) {
+        for (size_t i = 0; i < paths_num; i++) {
+            if (is_directory(paths[i])) {
+                load_directory(browser, paths[i]);
+            } else {
+                add_file(browser, paths[i]);
+            }
+        }
+    } else {
+        char cwd[PATH_MAX];
+        if (!cwd) {
+            fprintf(stderr, "Unable to get current directory: \n", strerror(errno));
+            return NULL;
+        }
+        getcwd(cwd, PATH_MAX);
+        load_directory(browser, cwd);
+    }
+    return browser;
+}
+
+const char* next_file(browser* context, bool forward)
+{
+    if (context->total == 0) {
+        return NULL;
+    }
+    int initial = context->current;
+    const int delta = forward ? 1 : -1;
+    do {
+        context->current += delta;
+        if (context->current == context->total) {
+            context->current = 0;
+        } else if (context->current < 0) {
+            context->current = context->total - 1;
+        }
+        if (context->current == initial) {
+            // we have looped around all files without finding anything
+            return NULL;
+        }
+    } while (!context->files[context->current]);
+    return context->files[context->current];
+}
+
+const char* current_file(browser* context)
+{
+    return context->files[context->current];
+}
+
+void delete_current_file(browser* context)
```

Comment:
You're right, `skip_current_file` is better name. But I have to disagree about the rest:
1. It is not meant to skip files which are not accessible (e.g. on network mounted volume), which would be quite rare, but rather to filter out non-images and incompatible images, which is quite common.
2. It is actually very common, for example I have many video files between my photos, which can't be showed in swayimg. Also various metadata, archives, photoshop files, etc.
3. Yes, I did test that multiple incompatible files and `next_file` skips them correctly. For all incompatible files, the function returns `NULL` and program exits after printing "No supported files to view". I can write proper unit tests if you want.

Reply:
Oh, I forgot to mention: Without this function, we would report incompatible files on every pass. I think it is better to report each error only once. And also not to waste time by opening files which we already know to not work on each iteration (if user cycles around, or goes backwards).

Reply:
Yeah, I forgot that this list contains list of ALL files, not only images. Too many garbage. May be we can use something like a filter by file extensions. What do you think?

Reply:
I was thinking about extension filter too, but I'm afraid it would be rather error-prone. Even such common format as JPEG has at least [6 standard extensions](https://en.wikipedia.org/wiki/JPEG#JPEG_filename_extensions), and I'm pretty sure there is many others not listed in that article. Also, we would have to normalize it to lowercase, for example my old Cannon camera stores everything in uppercase for some reason. Not to mention that the very same camera stores  thumbnails of video files in .THM files, which are valid JPEG (and swayimage currently shows them without problem).

I think that such filter would be good as a commandline option, so the user can specify which files he desires to view. It might actually be a glob or regexp of full file name. But having a built-in extension filter seems too constraining to me.

Unfortunately, as far as I can tell, the only 100% sure way to determine, whether file is really image is to try to load it. But I think it might be possible to actually do just that. We could fill the `files`structure lazily, if we keep the pointer for the FTS* struct in the `browser`, than we can iterate over it as user steps through the files. It would require some architectural changes (again :slightly_smiling_face:), but in the end, files would be just a "cache" of valid files, while the first run of forward iteration would run directly on the `fts_children` structure.

By the way: I think that the results of fts_* functions can take quite a lot of memory by itself, since it stores all the info about the files (not only path, but also inode, `stat()` result, etc.). We could probably save quite a lot memory by using something simpler (like `opendir`+`readdir`+`closedir`), but it would mean we have to take care of the recursion and sorting ourselves. I'm honestly not sure which way is better.

Reply:
I tried to measure the memory footprint... It seems that recursively loading my photo folder with about 150 directories containing some 9000 photos takes about 3MB in resident memory. Not great, but I think I could live with that.